chainer/optimizer.py

```diff
@@ -357,33 +534,40 @@ class GradientMethod(Optimizer):
     methods that just require the gradient at the current parameter vector on
     an update can be implemented as its child class.
 
-    An implementation of a gradient method must override the following methods:
-
-    - :meth:`init_state` or both :meth:`init_state_cpu` and
-      :meth:`init_state_gpu`
-    - :meth:`update_one` or both :meth:`update_one_cpu` and
-      :meth:`update_one_gpu`
+    This class uses :class:`~chainer.UpdateRule` to manage the update rule of
+    each parameter. A child class of GradientMethod should override
+    :meth:`setup_update_rule` to set up the default update rule to each
+    parameter.
 
     .. note::
        It is recommended to call :meth:`use_cleargrads` after creating a
        :class:`GradientMethod` object for efficiency.
 
     """
+    def __init__(self):
+        super(GradientMethod, self).__init__()
+        self.hyperparam = Hyperparameter()
+
+    def setup(self, link):
+        super(GradientMethod, self).setup(link)
+        for param in link.params():
+            rule = self.create_update_rule()
+            rule.hyperparam.set_parent(self.hyperparam)
+            param.update_rule = rule
```

Comment:
I think it's an unnecessary coupling between `Optimizer` and `UpdateRule` to set parent hyperparameter here.
How about setting it via `UpdateRule`'s constructor, like this?

```
class UpdateRule:
    def __init__(self, parent_hyperparam=None):
        self.hyperparam = Hyperparameter(parent_hyperparam)
```

```
class AdamRule:
    def __init__(self, parent_hyperparam=None, ...):
        super().__init__(parent_hyperparam or _default_hyperparam)
```

```
class Adam:
    def create_update_rule(self):
        return AdamRule(self.parent_hyperparam)
```

Then `Hyperparameter.set_parent()` could be no longer needed.